File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.append(-1) # Guarantee max area will be computed at last index
        max_area, stack = float('-inf'), []
        
        for i, h in enumerate(heights):
            index = i
            while stack and stack[-1][1] > h:
                index, height = stack.pop()
                max_area = max(max_area, height * (i - index))
            stack.append((index,h))
                
                
        return max_area

File: 84-largest-rectangle-in-histogram/84-largest-rectangle-in-histogram.ts
function largestRectangleArea(heights) {
    const stack = [];
    
    let area = 0;
    let max = 0;
    
    heights.push(0);
    
    for (const height of heights) {
        
        let multiplier = 0;
        
        while (stack.length > 0 && height < stack[stack.length - 1][0]) {
            const tallest = stack.pop();
            multiplier += 1 + tallest[1];
            const area = multiplier * tallest[0];
            
            if (area > max) {
                max = area;
            }
        }
        
        stack.push([height, multiplier]);
        
    }
    
    return max;
}